File: mirrorbooth/android/app/src/main/kotlin/com/ol1n/mirrorbooth/ShutterPlugin.kt
package com.ol1n.mirrorbooth

import android.os.SystemClock
import android.view.KeyEvent
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Volume-button shutter. MainActivity forwards key events here; while enabled
 * the volume keys are consumed (no system volume change) and reported to Dart
 * as short or long presses.
 */
class ShutterPlugin :
    FlutterPlugin, MethodChannel.MethodCallHandler, EventChannel.StreamHandler {

    private companion object {
        /** Hold this long to toggle recording instead of taking a photo. */
        const val LONG_PRESS_MS = 500L
    }

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null

    @Volatile
    private var enabled = false
    private var pressStart: Long? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel = MethodChannel(binding.binaryMessenger, "mirrorbooth/shutter")
        methodChannel.setMethodCallHandler(this)
        eventChannel = EventChannel(binding.binaryMessenger, "mirrorbooth/shutter_events")
        eventChannel.setStreamHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "setEnabled" -> {
                enabled = call.argument<Boolean>("enabled") ?: false
                if (!enabled) pressStart = null
                result.success(null)
            }
            "isSupported" -> result.success(true)
            else -> result.notImplemented()
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        eventSink = null
    }

    /** Returns true when the event was consumed as a shutter press. */
    fun handleKeyEvent(event: KeyEvent): Boolean {
        if (!enabled) return false
        if (event.keyCode != KeyEvent.KEYCODE_VOLUME_UP &&
            event.keyCode != KeyEvent.KEYCODE_VOLUME_DOWN
        ) {
            return false
        }
        when (event.action) {
            KeyEvent.ACTION_DOWN -> if (event.repeatCount == 0) {
                pressStart = SystemClock.uptimeMillis()
            }
            KeyEvent.ACTION_UP -> {
                val start = pressStart
                pressStart = null
                if (start != null) {
                    val held = SystemClock.uptimeMillis() - start
                    eventSink?.success(mapOf("long" to (held >= LONG_PRESS_MS)))
                }
            }
        }
        return true
    }
}

File: mirrorbooth/ios/Runner/ShutterPlugin.swift
import AVKit
import Flutter
import UIKit

/// Hardware capture-button shutter (volume buttons, Action button, Camera
/// Control) via AVCaptureEventInteraction — the public API Apple provides for
/// camera apps on iOS 17.2+. Older systems simply never emit events and the
/// on-screen buttons remain the only trigger.
public class ShutterPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    /// Hold this long to toggle recording instead of taking a photo.
    private static let longPressSeconds: CFTimeInterval = 0.5

    private var eventSink: FlutterEventSink?
    private var pressStart: CFTimeInterval?
    /// Typed as Any? so the class still compiles/loads below iOS 17.2.
    private var interaction: Any?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = ShutterPlugin()
        let methodChannel = FlutterMethodChannel(
            name: "mirrorbooth/shutter",
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        let eventChannel = FlutterEventChannel(
            name: "mirrorbooth/shutter_events",
            binaryMessenger: registrar.messenger()
        )
        eventChannel.setStreamHandler(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "setEnabled":
            let args = call.arguments as? [String: Any]
            setEnabled((args?["enabled"] as? Bool) ?? false)
            result(nil)
        case "isSupported":
            if #available(iOS 17.2, *) { result(true) } else { result(false) }
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - FlutterStreamHandler

    public func onListen(
        withArguments arguments: Any?,
        eventSink events: @escaping FlutterEventSink
    ) -> FlutterError? {
        eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }

    // MARK: - Capture events

    private func setEnabled(_ enabled: Bool) {
        guard #available(iOS 17.2, *) else { return }
        if enabled {
            guard interaction == nil, let view = rootView() else { return }
            let captureInteraction = AVCaptureEventInteraction { [weak self] event in
                self?.handleCaptureEvent(event)
            }
            captureInteraction.isEnabled = true
            view.addInteraction(captureInteraction)
            interaction = captureInteraction
        } else {
            if let captureInteraction = interaction as? AVCaptureEventInteraction {
                captureInteraction.view?.removeInteraction(captureInteraction)
            }
            interaction = nil
            pressStart = nil
        }
    }

    @available(iOS 17.2, *)
    private func handleCaptureEvent(_ event: AVCaptureEvent) {
        switch event.phase {
        case .began:
            pressStart = CACurrentMediaTime()
        case .ended:
            let start = pressStart ?? CACurrentMediaTime()
            pressStart = nil
            emit(long: CACurrentMediaTime() - start >= Self.longPressSeconds)
        case .cancelled:
            pressStart = nil
        @unknown default:
            pressStart = nil
        }
    }

    private func emit(long: Bool) {
        guard let sink = eventSink else { return }
        if Thread.isMainThread {
            sink(["long": long])
        } else {
            DispatchQueue.main.async { sink(["long": long]) }
        }
    }

    private func rootView() -> UIView? {
        let windows = UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .flatMap { $0.windows }
        let window = windows.first { $0.isKeyWindow } ?? windows.first
        return window?.rootViewController?.view
    }
}
